homeassistant/components/picnic/sensor.py

```diff
@@ -0,0 +1,114 @@
+"""Definition of Picnic sensors."""
+
+from typing import Any, Optional
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ATTR_ATTRIBUTION
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import PlatformNotReady
+from homeassistant.helpers.typing import StateType
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+)
+
+from .const import ADDRESS, ATTRIBUTION, CONF_COORDINATOR, DOMAIN, SENSOR_TYPES
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, config_entry: ConfigEntry, async_add_entities
+):
+    """Set up Picnic sensor entries."""
+    picnic_coordinator = hass.data[DOMAIN][config_entry.entry_id][CONF_COORDINATOR]
+
+    # Fetch initial data so we have data when entities subscribe
+    await picnic_coordinator.async_refresh()
+    if not picnic_coordinator.last_update_success:
+        raise PlatformNotReady()
+
+    # Add an entity for each sensor type
+    async_add_entities(
+        PicnicSensor(picnic_coordinator, config_entry, sensor_type, props)
+        for sensor_type, props in SENSOR_TYPES.items()
+    )
+
+    return True
+
+
+class PicnicSensor(CoordinatorEntity):
+    """The CoordinatorEntity subclass representing Picnic sensors."""
+
+    def __init__(
+        self,
+        coordinator: DataUpdateCoordinator[Any],
+        config_entry: ConfigEntry,
+        sensor_type,
+        properties,
+    ):
+        """Init a Picnic sensor."""
+        super().__init__(coordinator)
+
+        self.sensor_type = sensor_type
+        self.properties = properties
+        self.entity_id = f"sensor.picnic_{sensor_type}"
+        self._service_unique_id = config_entry.unique_id
+
+    @property
+    def unit_of_measurement(self) -> Optional[str]:
+        """Return the unit this state is expressed in."""
+        return self.properties["unit"]
```

Comment:
If you change this to using `.get`, then they become optional in the source dict, allowing you to clean up `SENSOR_TYPES`.

```suggestion
        return self.properties.get("unit")
```